--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -150,6 +150,7 @@ ELSE()
     # This part is unnecessary 'casue the same is set by the lemon/core.h.
     # Still keep it as an example.
     SET(CXX_WARNING "/wd4250 /wd4355 /wd4503 /wd4800 /wd4996")
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /bigobj")
     # Suppressed warnings:
     # C4250: 'class1' : inherits 'class2::member' via dominance
     # C4355: 'this' : used in base member initializer list
