--- dlib/CMakeLists.txt	2015-06-03 20:15:45.000000000 -0400
+++ dlib/CMakeLists.txt	2015-06-29 14:11:44.000000000 -0400
@@ -398,6 +398,17 @@
    endif ()  ##### end of if NOT DLIB_ISO_CPP_ONLY ##########################################################
 
 
+   # Install headers
+   install(DIRECTORY ${PROJECT_SOURCE_DIR} 
+           DESTINATION include 
+           FILES_MATCHING PATTERN "*.h")
+
+   # Install lib
+   install(TARGETS dlib
+           RUNTIME DESTINATION bin 
+           LIBRARY DESTINATION lib 
+           ARCHIVE DESTINATION lib)
+
    #test for some things that really should be true about the compiler
    include(TestForSTDNamespace)
    include(TestForANSIStreamHeaders)
