diff --git numpy/distutils/command/build_ext.py numpy/distutils/command/build_ext.py
--- numpy/distutils/command/build_ext.py
+++ numpy/distutils/command/build_ext.py
@@ -456,7 +456,9 @@ class build_ext (old_build_ext):
             if libname.startswith('msvc'): continue
             fileexists = False
             for libdir in c_library_dirs or []:
-                libfile = os.path.join(libdir, '%s.lib' % (libname))
+                libfile = os.path.join(libdir,'lib%s.dll.a' % (libname))
+                if not os.path.isfile(libfile):
+                    libfile = os.path.join(libdir,'lib%s.a' % (libname))
                 if os.path.isfile(libfile):
                     fileexists = True
                     break
@@ -494,7 +496,9 @@ class build_ext (old_build_ext):
         for lib in fcompiler.libraries:
             if not lib.startswith('msvc'):
                 c_libraries.append(lib)
-                p = combine_paths(f_lib_dirs, 'lib' + lib + '.a')
+                p = combine_paths(f_lib_dirs, 'lib' + lib + '.dll.a')
+                if not p:
+                    p = combine_paths(f_lib_dirs, 'lib' + lib + '.a')
                 if p:
                     dst_name = os.path.join(self.build_temp, lib + '.lib')
                     if not os.path.isfile(dst_name):
diff --git numpy/distutils/fcompiler/gnu.py numpy/distutils/fcompiler/gnu.py
--- numpy/distutils/fcompiler/gnu.py
+++ numpy/distutils/fcompiler/gnu.py
@@ -62,7 +62,7 @@ class GnuFCompiler(FCompiler):
             m = re.search(r'GNU Fortran\s+95.*?([0-9-.]+)', version_string)
             if m:
                 return ('gfortran', m.group(1))
-            m = re.search(r'GNU Fortran.*?\-?([0-9-.]+)', version_string)
+            m = re.search(r'GNU Fortran.*?([0-9]+[-.][0-9]+[-.][0-9]+)', version_string)
             if m:
                 v = m.group(1)
                 if v.startswith('0') or v.startswith('2') or v.startswith('3'):
@@ -344,10 +344,7 @@ class Gnu95FCompiler(GnuFCompiler):
             # XXX: fix this mess, does not work for mingw
             if is_win64():
                 c_compiler = self.c_compiler
-                if c_compiler and c_compiler.compiler_type == "msvc":
-                    return []
-                else:
-                    pass
+                return []
         return opt

     def get_target(self):
diff --git numpy/distutils/misc_util.py numpy/distutils/misc_util.py
--- numpy/distutils/misc_util.py
+++ numpy/distutils/misc_util.py
@@ -397,6 +397,7 @@ def msvc_runtime_library():
                '1400': 'msvcr80',    # MSVC 8
                '1500': 'msvcr90',    # MSVC 9 (VS 2008)
                '1600': 'msvcr100',   # MSVC 10 (aka 2010)
+               '1700': 'msvcr110',   # MSVC 11 (aka 2012)
               }.get(msc_ver, None)
     else:
         lib = None
diff --git numpy/distutils/msvccompiler.py numpy/distutils/msvccompiler.py
--- numpy/distutils/msvccompiler.py
+++ numpy/distutils/msvccompiler.py
@@ -8,6 +8,7 @@ from .system_info import platform_bits
 class MSVCCompiler(distutils.msvccompiler.MSVCCompiler):
     def __init__(self, verbose=0, dry_run=0, force=0):
         distutils.msvccompiler.MSVCCompiler.__init__(self, verbose, dry_run, force)
+        self.compiler_cxx = ['cl.exe']

     def initialize(self, plat_name=None):
         environ_lib = os.getenv('lib')
